### PR TITLE
Add __stream_map__, __stream_reduce__, and __stream_merge__ protocols

### PR DESCRIPTION
**Pull request by mrocklin**
This allows objects within a stream to take control of map, accumulate, and
combination operations (like zip, combine_latest).  I *hope* that this helps to
resolve problems around dask, metadata tracking, batching, etc..

cc @ordirules @danielballan

### CODE REVIEW

File: streams/core.py [hunk 1]
```diff
@@ -474,6 +483,8 @@ def update(self, x, who=None):
         if len(L) == 1 and all(self.buffers):
             tup = tuple(buf.popleft() for buf in self.buffers)
             self.condition.notify_all()
+            if tup and hasattr(tup[0], '__stream_merge__'):
+                tup = tup[0].__stream_merge__(*tup[1:])
```
Comment (mrocklin) [inline streams/core.py]: In the future this would need some sort of dispatching mechanism to combine different kinds of objects.

**Comment by ordirules** [comment]
I just found it will take a bit of work to get to function with `Delayed`. The reason is that I now need to subclass `Delayed` (adding the `__stream_map__` etc methods). My previous wrapper method allowed me to avoid this by just supplying the `delayed` routine as a wrapper.

Basically, as the data comes in, I need to transform it with `delayed(wrapper(func))(data)` where `data` is encoded/decoded by `wrapper`. Adding this as a wrapper to streams was enough previously to get this working, and running on cluster.

Any tips on getting that implemented? Would it be simple enough to subclass `Delayed` for this? I currently can't see a simple way, but I'm not too familiar with the library. Thanks!

**Comment by mrocklin** [comment]
I'm somewhat curious why you're passing delayed objects through the stream.  

We might consider using something like `functools.singledispatch` instead of a `__stream_map__` attribute.

**Comment by mrocklin** [comment]
I started making a batched solution with the protocol and found that I also needed to change some things around.  I suspect that we'll need to iterate on this a few times before it's stable.

**Comment by ordirules** [comment]
I'm passing delayed objects through the stream to construct a task graph of delayed computations. I can then choose to submit pieces of the stream to the cluster if needed and save the futures to deques, to cache various intermediate steps.
If there is a cleaner solution, I would be interested to hear. Currently this has worked in the past.

I'm not sure singledispatch would work. For example, what if these wrappers were nested? How would you distinguish them wrapper1(wrapper2(f)) from wrapper1(wrapper3(f))? I'm thinking of the scenario where a user may have their own wrapper that they `delay` (which again may not be useful). However, this may not be a common usage case.

**Comment by mrocklin** [comment]
I've also been running into the nesting problem.  (futures of batches).

**Comment by mrocklin** [comment]
I plan to work on this a bit today.  My plan is the following:

1.  Implement a Batch type that maps over many elements.  This will require me to change some of the protocols like `__stream_reduce__` and `__stream_merge__`
2.  Try to handle Futures of Batches.  I think that this will force me to think about nesting.
3.  Remove the monkey-patching of Futures, which I think will require the use of a dispatching solution

I think that this will cover the cases that you bring up.  Regardless it would be useful to have your input if I'm able to put something together today.

**Comment by ordirules** [comment]
That would be great. I have been a little busy on my end, but I will definitely try to read and offer input and test things out.

The nesting piece will be very tricky I think. I think it seems like there are two choices: either you give the stream information on how to handle the data, or else you force the data to know about it. You mentioned you don't want to do the former to keep the API clean, if I understand. I agree. The fact this library is lightweight is a nice plus.

I think for my specific case, subclassing a Delayed object to evaluate certain `hasattr` checks could work, so that:
```python
hasattr(x, '__stream_map__')
```
would evaluate true for a delayed(StreamDoc(something)), and so the object would know to do the mapping. It might be hackish though. But I think delayed is a special type of class and so it probably needs to be handled differently. I haven't tried this since I don't know so much about delayed to see if it's possible. If you had input, I'd be happy to hear.

Also, forcing user to write wrappers that are aware that they're nesting other wrappers is probably not a bad constraint (so the object with the `__stream_wrapper__` method must check for the next returned doc's `__stream_wrapper__`). If you're nesting, you should be aware of what you're doing.

Please correct me if I misunderstood anything. I hope this gives enough user input. By the way, I have reverted to using the wrapper before temporarily (since I can't delay objects so easily). It was pretty easy to switch to the new method, so I'm confident that I can update our library to any changes to test them. This allows me to test more features while we wring out these details. Thanks again for all the support!

**Pull request merged**